File: src/lib/numberToWords.ts
export const numberToWords = (num: number | string): string => {
  const n = parseInt(num.toString(), 10);
  if (isNaN(n) || n === 0) return 'ZERO';
  
  const a = ['', 'ONE ', 'TWO ', 'THREE ', 'FOUR ', 'FIVE ', 'SIX ', 'SEVEN ', 'EIGHT ', 'NINE ', 'TEN ', 'ELEVEN ', 'TWELVE ', 'THIRTEEN ', 'FOURTEEN ', 'FIFTEEN ', 'SIXTEEN ', 'SEVENTEEN ', 'EIGHTEEN ', 'NINETEEN '];
  const b = ['', '', 'TWENTY', 'THIRTY', 'FORTY', 'FIFTY', 'SIXTY', 'SEVENTY', 'EIGHTY', 'NINETY'];

  const numStr = n.toString();
  if (numStr.length > 9) return 'OVERFLOW';
  
  const match = ('000000000' + numStr).substr(-9).match(/^(\d{2})(\d{2})(\d{2})(\d{1})(\d{2})$/);
  if (!match) return '';
  
  let str = '';
  str += (match[1] != '00') ? (a[Number(match[1])] || b[match[1][0] as any] + ' ' + a[match[1][1] as any]) + 'CRORE ' : '';
  str += (match[2] != '00') ? (a[Number(match[2])] || b[match[2][0] as any] + ' ' + a[match[2][1] as any]) + 'LAKH ' : '';
  str += (match[3] != '00') ? (a[Number(match[3])] || b[match[3][0] as any] + ' ' + a[match[3][1] as any]) + 'THOUSAND ' : '';
  str += (match[4] != '0') ? (a[Number(match[4])] || b[match[4][0] as any] + ' ' + a[match[4][1] as any]) + 'HUNDRED ' : '';
  str += (match[5] != '00') ? ((str != '') ? 'AND ' : '') + (a[Number(match[5])] || b[match[5][0] as any] + ' ' + a[match[5][1] as any]) : '';
  
  return str.trim();
};

File: android/app/src/main/java/com/billcraft/studio/utils/NumberToWords.java
package com.billcraft.studio.utils;

public class NumberToWords {
    private static final String[] units = {
        "", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten",
        "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"
    };

    private static final String[] tens = {
        "", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"
    };

    public static String convertToIndianRupees(double amount) {
        long wholePart = (long) amount;
        int paise = (int) Math.round((amount - wholePart) * 100);

        if (wholePart == 0 && paise == 0) return "Zero Rupees Only";

        StringBuilder sb = new StringBuilder();
        if (wholePart > 0) {
            sb.append(convertNumber(wholePart)).append(" Rupees");
        }
        if (paise > 0) {
            if (sb.length() > 0) sb.append(" and ");
            sb.append(convertNumber(paise)).append(" Paise");
        }
        sb.append(" Only");
        return sb.toString();
    }

    private static String convertNumber(long n) {
        if (n < 0) return "Minus " + convertNumber(-n);
        if (n <= 19) return units[(int) n];
        if (n <= 99) return tens[(int) (n / 10)] + (n % 10 != 0 ? " " + units[(int) (n % 10)] : "");
        if (n <= 999) return convertNumber(n / 100) + " Hundred" + (n % 100 != 0 ? " " + convertNumber(n % 100) : "");
        if (n <= 99999) return convertNumber(n / 1000) + " Thousand" + (n % 1000 != 0 ? " " + convertNumber(n % 1000) : "");
        if (n <= 9999999) return convertNumber(n / 100000) + " Lakh" + (n % 100000 != 0 ? " " + convertNumber(n % 100000) : "");
        return convertNumber(n / 10000000) + " Crore" + (n % 10000000 != 0 ? " " + convertNumber(n % 10000000) : "");
    }
}
